### PR TITLE
Release notes for 0.22.0

### PR DESCRIPTION
**Pull request by ryanking13**
This is a draft of 0.22.0 release note. Feel free to modify, remove, or add stuff.

I think the most remarkable part of this release is FFI improvements by Hood,
But I wasn't able to summarize the whole FFI improvements properly, so I would appriciate if @hoodmane can revise.

_Disclosure:  I wrote this draft and asked ChatGPT to revise 😂_

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 public/
+.hugo_build.lock
```
Comment (hoodmane) [inline .gitignore:2]:
```suggestion
.hugo_build.lock

```

File: content/posts/0.22-release.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+---
+
+Pyodide 0.22 is a major release focused on improving the
+foreign function interface (FFI) between Python and JavaScript.
+The goal of Pyodide is to make it easy to use JavaScript objects
+in Python as if they were native Python objects, and vice versa.
+In this release, we have made several improvements to the FFI
+to achieve this goal.
+
+We have also separated micropip, the lightweight package manager used in Pyodide,
```
Comment (hoodmane) [inline content/posts/0.22-release.md:27]: > separate

I think the term "unvendor" would probably be better here.

File: content/posts/0.22-release.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+on JavaScript map-like objects from Python.
+
+TODO: PyProxy?
+TODO: add example?
+
+## Separating packages in Pyodide
+
+From the previous 0.21.0 release, we have been separating certain packages
+that are used in Pyodide into standalone repositories.
+These packages are now available as standalone repositories:
```
Comment (phorward) [inline content/posts/0.22-release.md:58]:
```suggestion
From the previous 0.21.0 release, we have been separating certain packages
that are used in Pyodide into standalone repositories.

These packages are now available as standalone repositories:
```

**Comment by hoodmane** [comment]
I will make some adjustments and commit here.

**Comment by hoodmane** [comment]
Thanks @ryanking13! I will try to fill in the ffi section soon.

> Disclosure: I wrote this draft and asked ChatGPT to revise

Cool!

File: content/posts/0.22-release.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+a pytest plugin for testing dedicated Pyodide applications.
+- [micropip](https://github.com/pyodide/micropip):
+a lightweight package manager for Pyodide and Python in the browser.
+- [matplotlib-pyodide](https://github.com/pyodide/matplotlib-pyodide):
+a matplotlib HTML backend.
+
+By separating these packages from the main Pyodide repository, we hope to
+facilitate their development and make it easier for the Python-in-the-browser
+community to use and improve them. They depend on the Pyodide ffi modules so for
+now it is still difficult to use them separate from Pyodide. In the future we
```
Comment (ryanking13) [inline content/posts/0.22-release.md:65]: Good point!

File: content/posts/0.22-release.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+hope to also unvendor the foreign function interface so that other Emscripten
+builds of CPython can use it.
+
+## Pyodide virtual environments
```
Comment (ryanking13) [inline content/posts/0.22-release.md:69]: Right, I forgot this one.

**Review by ryanking13 — COMMENTED** [review]
Thanks @hoodmane!

**Comment by rth** [comment]
Thanks, it's a nice read! Do we want to mention the `pyodide` CLI entry-point on the usability side?

**Comment by hoodmane** [comment]
> mention the pyodide CLI entry-point on the usability side?

Sounds good to me, but one of you should probably add it since I haven't used it much yet and don't know what is best to say about it.

File: content/posts/0.22-release.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+searchHidden: true
+ShowReadingTime: true
+ShowBreadCrumbs: true
+ShowPostNavLinks: true
+---
+
+Pyodide 0.22 is a major release focused on improving the foreign function
+interface between Python and JavaScript. We also improved the systems for
+building, loading, and testing packages. In particular, we added support for
+creating virtual environments that execute code using node Pyodide.
```
Comment (ryanking13) [inline content/posts/0.22-release.md:23]: `node Pyodide` doesn't seem very clear to me. Maybe something like `Pyodide and Node.js` instead?

File: content/posts/0.22-release.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+For a complete list of changes, please see the
+[changelog](https://pyodide.org/en/stable/project/changelog.html#version-0-22-0).
+
+## New JavaScript APIs
+
+We added support for mounting a folder from the local filesystem into the
+Pyodide filesystem using the Chrome [File System Access
+API](https://developer.chrome.com/articles/file-system-access/). If you acquire
+a directory handle (say from
+[showDirectoryPicker](https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker)) then you can mount this directory handle
```
Comment (ryanking13) [inline content/posts/0.22-release.md:41]:
```suggestion
[showDirectoryPicker](https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker)) then you can mount this directory handle.
```

File: content/posts/0.22-release.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+a directory handle (say from
+[showDirectoryPicker](https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker)) then you can mount this directory handle 
+
+We added new API to set the standard stream handlers from JavaScript and we
+improved the defaults for stream handling. The new stream APIs are particularly
+useful if you want to make a terminal-like interface to Python.
+
+## Pyodide virtual environments
+
+We added support for creating virtual environments that run Python using node
```
Comment (ryanking13) [inline content/posts/0.22-release.md:49]: Same here (node Pyodide)

**Review by ryanking13 — COMMENTED** [review]
Thanks for the revision @hoodmane. It looks very nice!

> Sounds good to me, but one of you should probably add it since I haven't used it much yet and don't know what is best to say about it.

I'll add a short introduction about it.

File: content/posts/0.22-release.md (hunk 6 above)
Comment (hoodmane) [reply content/posts/0.22-release.md:23]: Agreed that the phrasing is not good. Maybe "Pyodide in node.js"? How do we say it in the dosc?

File: content/posts/0.22-release.md [hunk 9]
```diff
@@ -155,6 +155,34 @@ let f = o.py_get_attr.bind(o);
 f("a"); // 3
 ```
 
+## The new Pyodide CLI interface
```
Comment (hoodmane) [inline content/posts/0.22-release.md:158]: My thought is to organize the sections roughly in order of increasing length, so maybe put this before the longer ffi section?

File: content/posts/0.22-release.md (hunk 6 above)
Comment (ryanking13) [reply content/posts/0.22-release.md:23]: "Pyodide in Node.js" sounds good to me. I wasn't able to find similar wording in docs though.

File: content/posts/0.22-release.md (hunk 9 above)
Comment (ryanking13) [reply content/posts/0.22-release.md:158]: > in order of increasing length

I would normally write in order of importance (things we want to emphasize), but in order of increasing length sounds also nice to me.

File: content/posts/0.22-release.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        self.a = a
+        self.b = b
+    
+    js_get_attr = run_js(
+        """
+        (function(name){
+            // `this` binds the `self` parameter of the method!
+            return this[name];
+        })
+        """
```
Comment (rth) [inline content/posts/0.22-release.md:133]: Yeah, it's pretty crazy that this works!

File: content/posts/0.22-release.md (hunk 6 above)
Comment (rth) [reply content/posts/0.22-release.md]: Fixed.

File: content/posts/0.22-release.md (hunk 9 above)
Comment (rth) [reply content/posts/0.22-release.md:158]: +1 also for the order of importance. Or I would say let's maybe just merge as is

**Review by rth — APPROVED** [review]
Thanks for working on this both of you!

I addressed the few minor remaining comments. +1 to merge, once the release is made.

**Comment by rth** [comment]
OK, I posted the release announcement with a link to these release notes on [Twitter](https://twitter.com/pyodide/status/1610303971507200002), [Mastodon](https://fosstodon.org/@pyodide) and [Linkedin](https://www.linkedin.com/company/pyodide/?viewAsMember=true). Thanks everyone.

**Comment by ryanking13** [comment]
Cool, I didn't know that we have a Mastodon account :)

**Pull request merged**